### DISCUSSION TITLE
How to preview images with imgcat on iTerm2?

### DISCUSSION THREAD

**Discussion by glaucocustodio** [junegunn/fzf#3609] [Q&A] [answered]
Hey guys, I have:

$ fzf --version
0.38.0 (brew)

$ imgcat -v
imgcat 2.5.1

and would like to know how should I set up FZF_DEFAULT_OPTS to make fzf preview images with iTerm2's imgcat, currently I've this: 

```
export FZF_DEFAULT_OPTS='
  --bind "ctrl-y:execute-silent(echo {+} | pbcopy)"
  --bind "ctrl-e:execute(echo {+} | xargs -o code)"
  --preview "([[ -f {} ]] && bat --theme=TwoDark --style=numbers --color=always --line-range :500 {})"
'
```

I've tried this but didn't work:

```
export FZF_DEFAULT_OPTS='
  --bind "ctrl-y:execute-silent(echo {+} | pbcopy)"
  --bind "ctrl-e:execute(echo {+} | xargs -o code)"
  --preview "
  if file --mime-type {} | grep -qF image/; then
    imgcat {}
  else
    bat --theme=TwoDark --style=numbers --color=always --line-range :500 {}
  fi
  "
'

```

**Comment by LangLangBart** [comment, marked as answer]
According to the changelog[^1], the `imgcat` functionality was introduced in version _0.44.0_.

Rather than using the iTerm2 `imgcat` script[^2], it appears you are using eddieantonio's `imgcat`[^3].

1. Consider updating to version _0.44.0_, and
2. modify the `FZF_DEFAULT_OPTS` as follows:
    - If you decide to use the iTerm2 `imgcat` script, please remove the `--depth iterm2` flag from the `FZF_DEFAULT_OPTS` below.

```bash
export FZF_DEFAULT_OPTS='
  --bind "ctrl-y:execute-silent(echo {+} | pbcopy)"
  --bind "ctrl-e:execute(echo {+} | xargs -o code)"
  --preview "
  if file --mime-type {} | grep -qF image/; then
    imgcat --depth iterm2 --width $FZF_PREVIEW_COLUMNS --height $FZF_PREVIEW_LINES {}
  else
    bat --theme=TwoDark --style=numbers --color=always --line-range :500 {}
  fi
  "
'
```

[^1]: [CHANGELOG.md 0.44.0](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0440)
[^2]: [iterm2.com/utilities/imgcat](https://iterm2.com/utilities/imgcat)
[^3]: [eddieantonio/imgcat](https://github.com/eddieantonio/imgcat)

**Reply by glaucocustodio** [reply]
thank you so much @LangLangBart, after updating brew (`brew update`) I was able to upgrade to fzf 0.46.1.

I tried iTerm2's imgcat but it didn't work, I ended up using [eddieantonio/imgcat](https://github.com/eddieantonio/imgcat) (which I installed with `brew install imgcat`) with FZF_DEFAULT_OPTS set as you posted

**Reply by LangLangBart** [reply]
>  [!NOTE]
> The maintainer recently pointed out in a comment that using the `--preview` flag is not recommended in the `FZF_DEFAULT_OPTS`.
>
> > The use of the --preview option in FZF_DEFAULT_OPTS is discouraged by me.
> > Source: [@junegunn · Issue #3608](https://github.com/junegunn/fzf/issues/3608#issuecomment-1924172099)

But, I have also included the `--preview` flag in my `FZF_DEFAULT_OPTS`. However, with a check to see if it's a file or a directory. If it's neither, for instance, a broken symlink file, there's a fallback solution.

This command was taken from the article: _Boost Your Command-Line Productivity With Fuzzy Finder_[^1]

```bash
--preview "([[ -f {} ]] && (bat --style=numbers --color=always {} || cat {})) ||
	([[ -d {} ]] && (tree -C {} | less)) ||
	echo {} 2> /dev/null | head -200"
```


<img src="https://github.com/junegunn/fzf/assets/92653266/c7248b70-05f0-43a0-9634-1d37d0209ca2" width="400">

[^1]: In 2020, this article was easily accessible. Now, you need to sign up for [Medium](https://betterprogramming.pub/boost-your-command-line-productivity-with-fuzzy-finder-985aa162ba5d)  or use [web.archive.org](https://web.archive.org/web/20211216190551/https://betterprogramming.pub/boost-your-command-line-productivity-with-fuzzy-finder-985aa162ba5d?gi=8695b9ddaf04) to read it. 

---

PS:  @junegunn, is the `--preview` command from the article something you would consider in your `README.md` ? The command can be found in 100's of repos on Github - [github.com/search](https://github.com/search?q=%22%28%5B%5B+-f+%7B%7D+%5D%5D+%26%26+%28bat+--style%3Dnumbers+--color%3Dalways+%7B%7D+%7C%7C+cat+%7B%7D%29%29+%7C%7C+%28%5B%5B+-d+%7B%7D+%5D%5D+%26%26+%28tree+-C+%7B%7D+%7C+less%29%29+%7C%7C+echo+%7B%7D+2%3E+%2Fdev%2Fnull+%7C+head+-200%22&type=code).

**Reply by junegunn** [reply]
I still don't think it's a good idea. You end up wasting half the screen if you use fzf with a list of processes, kubernetes pods, git branches, shell commands, etc.

Also, `head -200` part is not needed anymore since https://github.com/junegunn/fzf/commit/faf68dbc5cc52201d0962f73baa5a049528b913c#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L595-L602.

**Reply by LangLangBart** [reply]
> You end up wasting half the screen if you use fzf with a list of processes, kubernetes pods, git branches, shell commands, etc.

Ah, I understand your reasoning now. When the preview flag is not set in the `FZF_DEFAULT_OPTS`, no preview window will be shown at all. It really comes down to personal preferences, getting used to this setup, and optionally tweaking it with the `--preview-window` flag, for example on small windows. It's good to know that there's no harm done when placing the `--preview` flag in the `FZF_DEFAULT_OPTS`.
